# 02 — SQL Analysis
**Project:** Telco Customer Churn Prediction  
**Author:** Udit Jadli  
**Dataset:** Maven Analytics — Telecom Customer Churn  

## Objective
Design a relational PostgreSQL schema, load both tables from CSV, 
and write analytical SQL queries to uncover churn patterns across 
customer segments, contract types, and services.

## 1. Imports
We begin by importing the libraries needed to connect Python to PostgreSQL and load our data.

In [1]:
import pandas as pd
import numpy as np
import os
import psycopg2                        # connects Python directly to PostgreSQL
from sqlalchemy import create_engine   # lets us load dataframes into SQL tables
import warnings

warnings.filterwarnings('ignore')

print("Libraries imported successfully!")

Libraries imported successfully!


## 2. Database Connection
We establish a connection to our PostgreSQL database using SQLAlchemy. 
This allows us to load pandas DataFrames directly into SQL tables with a single command.

In [2]:
# connection string format: postgresql://username:password@host:port/database_name
engine = create_engine('postgresql://postgres:postgres97@localhost:5432/telecom_churn')

# create_engine('postgresql://postgres:postgres97@localhost:5432/telecom_churn')
# This is called a connection string — think of it like a URL that tells Python exactly where the database is and how to log in. Breaking it down:

# postgresql:// — the type of database
# postgres — the username
# postgres97 — the password
# localhost — the database is on your own machine
# 5432 — the default PostgreSQL port
#telecom_churn — the database we created

# test the connection
with engine.connect() as conn:
    print("Connected to telecom_churn database successfully!")

Connected to telecom_churn database successfully!


## 3. Load Data
We reload the cleaned CSV files into pandas DataFrames, then load them 
into PostgreSQL as relational tables using SQLAlchemy.

In [3]:
# define file paths
DATA_RAW = os.path.join('..', 'data', 'raw')

# load both tables we need for SQL analysis
df_churn = pd.read_csv(os.path.join(DATA_RAW, 'telecom_customer_churn.csv'), encoding='latin-1')
df_zip   = pd.read_csv(os.path.join(DATA_RAW, 'telecom_zipcode_population.csv'), encoding='latin-1')

print(f"Churn table:    {df_churn.shape[0]:,} rows × {df_churn.shape[1]} columns")
print(f"Zip code table: {df_zip.shape[0]:,} rows × {df_zip.shape[1]} columns")

Churn table:    7,043 rows × 38 columns
Zip code table: 1,671 rows × 2 columns


## 4. Column Name Cleaning
We convert all column names to snake_case before loading into PostgreSQL.
Column names with spaces or special characters cause errors in SQL queries.

In [4]:
import re

# snake_case: lowercase with underscores — e.g. "Customer ID" → "customer_id"
# PostgreSQL doesn't allow spaces in column names
def to_snake_case(col_name):
    col = col_name.strip()                      # remove extra spaces around the name
    col = re.sub(r'\s+', '_', col)              # replace spaces with underscores first
    col = re.sub(r'[^a-zA-Z0-9_]', '', col)    # remove special characters but keep underscores
    return col.lower()                          # convert everything to lowercase

# reload the raw data first to reset the column names
df_churn = pd.read_csv(os.path.join(DATA_RAW, 'telecom_customer_churn.csv'), encoding='latin-1')
df_zip   = pd.read_csv(os.path.join(DATA_RAW, 'telecom_zipcode_population.csv'), encoding='latin-1')

# now apply snake_case
df_churn.columns = [to_snake_case(c) for c in df_churn.columns]
df_zip.columns   = [to_snake_case(c) for c in df_zip.columns]

print("Churn table columns after cleaning:")
print(list(df_churn.columns))
print("\nZip code table columns after cleaning:")
print(list(df_zip.columns))

Churn table columns after cleaning:
['customer_id', 'gender', 'age', 'married', 'number_of_dependents', 'city', 'zip_code', 'latitude', 'longitude', 'number_of_referrals', 'tenure_in_months', 'offer', 'phone_service', 'avg_monthly_long_distance_charges', 'multiple_lines', 'internet_service', 'internet_type', 'avg_monthly_gb_download', 'online_security', 'online_backup', 'device_protection_plan', 'premium_tech_support', 'streaming_tv', 'streaming_movies', 'streaming_music', 'unlimited_data', 'contract', 'paperless_billing', 'payment_method', 'monthly_charge', 'total_charges', 'total_refunds', 'total_extra_data_charges', 'total_long_distance_charges', 'total_revenue', 'customer_status', 'churn_category', 'churn_reason']

Zip code table columns after cleaning:
['zip_code', 'population']


## 5. Load Tables into PostgreSQL
We load both DataFrames into PostgreSQL using SQLAlchemy's to_sql() method.
if_exists='replace' means if the table already exists, it gets dropped and recreated —
useful during development when we need to reload data cleanly.

In [5]:
# load churn table into PostgreSQL
df_churn.to_sql(
    'customer_churn',    # table name in PostgreSQL
    engine,              # our database connection
    if_exists='replace', # drop and recreate if table already exists
    index=False          # don't write the pandas index as a column
)
print("✓ customer_churn table loaded successfully")

# load zip code table into PostgreSQL
df_zip.to_sql(
    'zipcode_population',
    engine,
    if_exists='replace',
    index=False
)
print("✓ zipcode_population table loaded successfully")

✓ customer_churn table loaded successfully
✓ zipcode_population table loaded successfully


## 6. Verify Tables in PostgreSQL
We verify that both tables were created successfully by checking the 
row counts directly from the database.

In [6]:
from sqlalchemy import text

# connect and verify both tables loaded correctly
with engine.connect() as conn:
    
    # check customer_churn row count
    result = conn.execute(text("SELECT COUNT(*) FROM customer_churn"))
    print(f"customer_churn rows:     {result.fetchone()[0]:,}")
    
    # check zipcode_population row count
    result = conn.execute(text("SELECT COUNT(*) FROM zipcode_population"))
    print(f"zipcode_population rows: {result.fetchone()[0]:,}")

customer_churn rows:     7,043
zipcode_population rows: 1,671


## 7. SQL Analysis
We run analytical queries directly against the PostgreSQL tables to uncover 
churn patterns across customer segments, contract types, and services.
Each query answers a specific business question.

### Query 1 — Overall Churn Rate
**Business question:** What percentage of customers churned?

In [7]:
query1 = """
SELECT 
    customer_status,
    COUNT(*) AS customer_count,
    -- window function: divides each group count by the total to get percentage
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 1) AS percentage
FROM customer_churn
GROUP BY customer_status
ORDER BY customer_count DESC;
"""

df_q1 = pd.read_sql(query1, engine)
print("Query 1 — Overall Churn Rate\n")
print(df_q1.to_string(index=False))

Query 1 — Overall Churn Rate

customer_status  customer_count  percentage
         Stayed            4720        67.0
        Churned            1869        26.5
         Joined             454         6.4


**Finding:** 26.5% of customers churned in Q2 2022. 
This is our baseline churn rate — any segment above 26.5% is higher risk than average.

### Query 2 — Churn Rate by Contract Type
**Business question:** Which contract type has the highest churn rate?

In [8]:
query2 = """
SELECT
    contract,
    COUNT(*) AS total_customers,
    SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) AS churned,
    -- calculate churn rate for each contract type
    ROUND(SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS churn_rate_pct
FROM customer_churn
GROUP BY contract
ORDER BY churn_rate_pct DESC;
"""

df_q2 = pd.read_sql(query2, engine)
print("Query 2 — Churn Rate by Contract Type\n")
print(df_q2.to_string(index=False))

Query 2 — Churn Rate by Contract Type

      contract  total_customers  churned  churn_rate_pct
Month-to-Month             3610     1655            45.8
      One Year             1550      166            10.7
      Two Year             1883       48             2.5


**Finding:** Contract type is the strongest churn signal in the dataset.
Month-to-Month customers churn at 45.8% — nearly 18x higher than Two Year customers (2.5%).
Locking customers into longer contracts is clearly the most effective retention strategy.

### Query 3 — Churn Rate by Internet Type
**Business question:** Does the type of internet service affect churn rate?

In [9]:
query3 = """
SELECT
    COALESCE(internet_type, 'No Internet') AS internet_type,
    COUNT(*) AS total_customers,
    SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) AS churned,
    -- churn rate per internet type
    ROUND(SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS churn_rate_pct
FROM customer_churn
GROUP BY internet_type
ORDER BY churn_rate_pct DESC;
"""

df_q3 = pd.read_sql(query3, engine)
print("Query 3 — Churn Rate by Internet Type\n")
print(df_q3.to_string(index=False))

Query 3 — Churn Rate by Internet Type

internet_type  total_customers  churned  churn_rate_pct
  Fiber Optic             3035     1236            40.7
        Cable              830      213            25.7
          DSL             1652      307            18.6
  No Internet             1526      113             7.4


**Finding:** Fiber Optic customers have the highest churn rate at 40.7% — 
despite being the premium service. This suggests a value-for-money problem 
with the Fiber Optic offering, likely driven by competitor pricing.

### Query 4 — Churn Rate by Tenure Group
**Business question:** Are newer customers more likely to churn than long-term customers?

In [10]:
query4 = """
SELECT
    -- group tenure into meaningful business segments
    CASE 
        WHEN tenure_in_months <= 12  THEN '0-12 months'
        WHEN tenure_in_months <= 24  THEN '13-24 months'
        WHEN tenure_in_months <= 48  THEN '25-48 months'
        ELSE '49+ months'
    END AS tenure_group,
    COUNT(*) AS total_customers,
    SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) AS churned,
    ROUND(SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS churn_rate_pct
FROM customer_churn
GROUP BY tenure_group
ORDER BY churn_rate_pct DESC;
"""

df_q4 = pd.read_sql(query4, engine)
print("Query 4 — Churn Rate by Tenure Group\n")
print(df_q4.to_string(index=False))

Query 4 — Churn Rate by Tenure Group

tenure_group  total_customers  churned  churn_rate_pct
 0-12 months             2186     1037            47.4
13-24 months             1024      294            28.7
25-48 months             1594      325            20.4
  49+ months             2239      213             9.5


**Finding:** Churn is heavily concentrated in the first 12 months (47.4%). 
The first year is the critical retention window — customers who survive past 
12 months become significantly more loyal, dropping to 9.5% churn after 49+ months.

### Query 5 — Churn Rate by Payment Method
**Business question:** Does how a customer pays affect their likelihood to churn?

In [11]:
query5 = """
SELECT
    payment_method,
    COUNT(*) AS total_customers,
    SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) AS churned,
    -- churn rate per payment method
    ROUND(SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS churn_rate_pct
FROM customer_churn
GROUP BY payment_method
ORDER BY churn_rate_pct DESC;
"""

df_q5 = pd.read_sql(query5, engine)
print("Query 5 — Churn Rate by Payment Method\n")
print(df_q5.to_string(index=False))

Query 5 — Churn Rate by Payment Method

 payment_method  total_customers  churned  churn_rate_pct
   Mailed Check              385      142            36.9
Bank Withdrawal             3909     1329            34.0
    Credit Card             2749      398            14.5


**Finding:** Credit Card customers churn at just 14.5% — less than half the rate 
of Mailed Check (36.9%) and Bank Withdrawal (34.0%) customers. 
Encouraging customers to pay by credit card could be a low-cost retention strategy.

### Query 6 — Top 10 Churn Reasons
**Business question:** What specific reasons do customers give for leaving?

In [12]:
query6 = """
SELECT
    churn_reason,
    COUNT(*) AS customer_count,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 1) AS percentage
FROM customer_churn
WHERE customer_status = 'Churned'
    AND churn_reason IS NOT NULL
GROUP BY churn_reason
ORDER BY customer_count DESC
LIMIT 10;
"""

df_q6 = pd.read_sql(query6, engine)
print("Query 6 — Top 10 Churn Reasons\n")
print(df_q6.to_string(index=False))

Query 6 — Top 10 Churn Reasons

                             churn_reason  customer_count  percentage
            Competitor had better devices             313        16.7
             Competitor made better offer             311        16.6
               Attitude of support person             220        11.8
                               Don't know             130         7.0
             Competitor offered more data             117         6.3
Competitor offered higher download speeds             100         5.4
             Attitude of service provider              94         5.0
                           Price too high              78         4.2
                  Product dissatisfaction              77         4.1
                      Network reliability              72         3.9


**Finding:** Competitor advantages drive 44%+ of specific churn reasons, 
with better devices and better offers leading. However, attitude of support 
staff (11.8%) is the top controllable factor — a service quality issue 
the company can directly address without changing pricing.

### Query 7 — Churn Rate by Offer Type
**Business question:** Do marketing offers reduce churn — and which ones work best?

In [13]:
query7 = """
SELECT
    COALESCE(offer, 'No Offer') AS offer,
    COUNT(*) AS total_customers,
    SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) AS churned,
    -- churn rate per offer type
    ROUND(SUM(CASE WHEN customer_status = 'Churned' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS churn_rate_pct
FROM customer_churn
GROUP BY offer
ORDER BY churn_rate_pct DESC;
"""

df_q7 = pd.read_sql(query7, engine)
print("Query 7 — Churn Rate by Offer Type\n")
print(df_q7.to_string(index=False))

Query 7 — Churn Rate by Offer Type

   offer  total_customers  churned  churn_rate_pct
 Offer E              805      426            52.9
No Offer             3877     1051            27.1
 Offer D              602      161            26.7
 Offer C              415       95            22.9
 Offer B              824      101            12.3
 Offer A              520       35             6.7


**Finding:** Not all offers reduce churn — Offer E (52.9%) is associated with 
higher churn than having no offer at all (27.1%). Offer A is the most effective 
retention tool at just 6.7% churn. The company should investigate why Offer E 
is underperforming and reallocate budget toward Offer A.

### Query 8 — Top 10 Cities by Revenue Lost to Churn
**Business question:** Which cities are generating the most revenue loss from churned customers?

In [14]:
query8 = """
SELECT
    city,
    COUNT(*) AS churned_customers,
    ROUND(AVG(monthly_charge)::numeric, 2) AS avg_monthly_charge,
    -- cast to numeric first — PostgreSQL requires explicit casting for ROUND with decimals
    ROUND(SUM(total_revenue)::numeric, 2) AS total_revenue_lost
FROM customer_churn
WHERE customer_status = 'Churned'
GROUP BY city
HAVING COUNT(*) >= 10
ORDER BY total_revenue_lost DESC
LIMIT 10;
"""

df_q8 = pd.read_sql(query8, engine)
print("Query 8 — Top 10 Cities by Revenue Lost to Churn\n")
print(df_q8.to_string(index=False))

Query 8 — Top 10 Cities by Revenue Lost to Churn

         city  churned_customers  avg_monthly_charge  total_revenue_lost
    San Diego                185               72.97           385446.39
  Los Angeles                 78               73.54           147090.46
   Sacramento                 26               72.74            59449.20
San Francisco                 31               79.89            56894.27
Santa Barbara                 10               86.07            46033.71
     San Jose                 29               70.05            44393.49
     Temecula                 22               81.18            44102.41
    Escondido                 16               80.09            38869.89
    Fallbrook                 26               70.21            34691.78
       Fresno                 13               67.97            29607.42


**Finding:** San Diego is the highest revenue loss city at $385,446 — 
nearly 3x more than Los Angeles ($147,090). San Francisco churned customers 
have the highest average monthly charge ($79.89), making each churned 
customer there more valuable to retain.

### Query 9 — Churn Rate by Population Size
**Business question:** Do customers in densely populated areas churn differently 
than those in smaller areas?

In [15]:
query9 = """
SELECT
    -- group zip codes into population buckets
    CASE
        WHEN z.population < 20000  THEN 'Small (< 20k)'
        WHEN z.population < 50000  THEN 'Medium (20k-50k)'
        WHEN z.population < 100000 THEN 'Large (50k-100k)'
        ELSE 'Very Large (100k+)'
    END AS population_group,
    COUNT(*) AS total_customers,
    SUM(CASE WHEN c.customer_status = 'Churned' THEN 1 ELSE 0 END) AS churned,
    ROUND(SUM(CASE WHEN c.customer_status = 'Churned' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS churn_rate_pct
FROM customer_churn c
-- JOIN the population table using zip_code as the key
JOIN zipcode_population z ON c.zip_code = z.zip_code
GROUP BY population_group
ORDER BY churn_rate_pct DESC;
"""

df_q9 = pd.read_sql(query9, engine)
print("Query 9 — Churn Rate by Population Size\n")
print(df_q9.to_string(index=False))

Query 9 — Churn Rate by Population Size

  population_group  total_customers  churned  churn_rate_pct
  Large (50k-100k)              713      219            30.7
  Medium (20k-50k)             2580      735            28.5
     Small (< 20k)             3735      914            24.5
Very Large (100k+)               15        1             6.7


**Finding:** Customers in larger populated areas (50k-100k) churn at 30.7% 
compared to 24.5% in smaller areas — likely because urban customers have 
more competitor options available. Note: the Very Large (100k+) group only 
has 15 customers and is not statistically reliable.

### Query 10 — High Value Customers at Risk
**Business question:** Which churned customers were the highest value — 
and what did they have in common?

In [16]:
query10 = """
SELECT
    customer_id,
    city,
    contract,
    internet_type,
    monthly_charge,
    total_revenue,
    churn_reason
FROM customer_churn
WHERE customer_status = 'Churned'
    -- focus on high value customers — top quartile monthly charge
    AND monthly_charge > 85
ORDER BY total_revenue DESC
LIMIT 10;
"""

df_q10 = pd.read_sql(query10, engine)
print("Query 10 — Top 10 High Value Churned Customers\n")
print(df_q10.to_string(index=False))

Query 10 — Top 10 High Value Churned Customers

customer_id          city       contract internet_type  monthly_charge  total_revenue                              churn_reason
 2889-FPWRM Mckinleyville       One Year   Fiber Optic          117.80       11195.44 Competitor offered higher download speeds
 3259-FDWOY     Elk Grove       Two Year   Fiber Optic          106.00       11084.84 Competitor offered higher download speeds
 2834-JRTUA      Martinez       Two Year   Fiber Optic          108.05       11040.97             Competitor had better devices
 1984-FCOWB     Templeton       One Year   Fiber Optic          109.50       10756.15              Attitude of service provider
 5287-QWLKY      Big Pine Month-to-Month   Fiber Optic          105.10       10718.96                Attitude of support person
 9835-ZIITK      Campbell       One Year   Fiber Optic          110.85       10717.17              Competitor made better offer
 5440-FLBQG Santa Barbara       Two Year   Fiber Optic  

**Finding:** Every top 10 high-value churned customer was on Fiber Optic, 
with total revenue losses of $10,000+ per customer. Competitor advantages 
(speed, devices, offers) are the primary drivers. Retaining even a fraction 
of these customers would have significant revenue impact.

## Summary of SQL Findings

| Query | Business Question | Key Finding |
|---|---|---|
| 1 | Overall churn rate | 26.5% baseline churn rate |
| 2 | Churn by contract | Month-to-Month 45.8% vs Two Year 2.5% |
| 3 | Churn by internet type | Fiber Optic highest at 40.7% |
| 4 | Churn by tenure | First 12 months critical — 47.4% churn |
| 5 | Churn by payment method | Credit Card lowest at 14.5% |
| 6 | Top churn reasons | Competitor advantages drive 44%+ of churn |
| 7 | Churn by offer type | Offer A best (6.7%), Offer E worst (52.9%) |
| 8 | Revenue lost by city | San Diego $385k — highest revenue loss |
| 9 | Churn by population | Larger areas churn slightly more |
| 10 | High value customers | All top churned customers on Fiber Optic |

## Next Step
`03_eda.ipynb` — deep exploratory data analysis with visualisations.